linker/resolve.go

```diff
@@ -34,11 +34,11 @@ func (r *result) ResolveMessageLiteralExtensionName(node ast.IdentValueNode) str
 	return r.optionQualifiedNames[node]
 }
 
-func (r *result) resolveElement(name protoreflect.FullName) protoreflect.Descriptor {
+func (r *result) resolveElement(name protoreflect.FullName, checkedCache []string) protoreflect.Descriptor {
```

Comment:
Have you tried a `map[string]struct{}` for the set here? Should be quicker than string comparison for arrays in most  cases I'd have thought.

Reply:
At FullStory, we found this was usually not the case: at least with earlier version of Go (since it's been a while since we had this exercise), map construction was _substantially_ slower than slice construction. And even though maps ostensibly offer O(1) lookup (assuming no/few collisions), the constant factors are _much_ higher than indexing into a slice such that the map had to be larger than you would guess for them to actually be faster. In this case, the maximum slice length will be the number of files in the transitive dependency graph for `r`, which is usually rather small for a single proto source file. The other downside to a map is that they take O(N) to clear/reset (at least to do so w/out allocations) whereas it's O(1) to do so with a slice.

I suspect there will be almost no measurable difference, and if there is a difference, it will be slices that are a wee bit faster. But I'll go ahead and try it and re-run some benchmarks to verify and report back to you. If they come out ahead, I'll happily change this.

Reply:
Neat, thanks for trying.